rfc/src/rfcs/0008-source-coverage.md

```diff
@@ -0,0 +1,277 @@
+Each `.profraw` file corresponds to a specific source code file in your project.
+
+At this point, we'll have produced the artifacts that we generally require for the LLVM tools:
+ 1. **The instrumented binary** which, in addition to the instrumented program,
+ contains additional information (e.g., the coverage mappings) required to
+ interpret the profiling results.
+ 2. **The `.profraw` files** which essentially includes the profiling results
+ (counter and expression values) for each function of the corresponding source
+ code file.
+
+For step 3, the commands will depend on what kind of results we want.
+Most likely we will have to merge the `.profraw` files and produce a `.profdata` file as follows:
+
+```sh
+llvm-profdata merge -sparse *.profraw -o output.profdata
+```
+
+Then, we can use a command such as
+
+```sh
+llvm-cov show target/debug/binary —instr-profile=output.profdata -show-line-counts-or-regions
+```
+
+to visualize the code coverage through the terminal as in the image:
+
+![Source-based code coverage with `llvm-cov`](https://github.com/model-checking/kani/assets/73246657/4f8a973d-8977-4c0b-822d-e73ed6d223aa)
+
+or the command
+
+```sh
+llvm-cov report target/debug/binary --instr-profile=output.profdata --show-region-summary
+```
+
+to produce coverage summaries like this:
+
+```
+Filename                                             Regions    Missed Regions     Cover   Functions  Missed Functions  Executed       Lines      Missed Lines     Cover    Branches   Missed Branches     Cover
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+/long/long/path/to/my/project/binary/src/main.rs           9                 3    66.67%           3                 1    66.67%          14                 4    71.43%           0                 0         -
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+TOTAL                                                      9                 3    66.67%           3                 1    66.67%          14                 4    71.43%           0                 0         -
+```
+
+[^note-source]: The LLVM project refers to their own coverage feature as "source-based code coverage".
+It's not rare to see the term "region coverage" being used instead to refer to the same thing.
+That's because LLVM's source-based code coverage feature can report coverage for code regions,
+but other coverage frameworks don't support the concept of code regions.
+
+### The Kani coverage workflow
+
+The proposed Kani coverage workflow imitates the LLVM coverage workflow as much as possible.
```

Comment:
This seems to imply that to verify the code you run `kani`, and to produce the coverage information you run `kani cov`. My guess is that both of these will need to run symex (though only the first one will run SAT, or at least the property-checking part of SAT).

This seems unnecessary. Why not make coverage an option to kani instead of a separate subcommand? That is, `kani --dump-coverage` means "verify the program, and while you're at it, dump out the `.kaniraw` and `.kanimap` that will later be postprocessed".

I agree that coverage checking will become a vital part of many users' workflow (as you've noted with the bullets beginning with "we expect users to employ this coverage-related option on several stages of a verification effort"). So I think it's important not to waste time (both compute time and running multiple commands). One thing I haven't seen mentioned is IDE integration; I would imagine users would want to press a single button and get both the verification results, and also coverage info highlighted in the IDE... it seems to me that with the proposed flow, the IDE would have to run kani twice.

Reply:
1. You're right, there's no need to have a subcommand. I haven't found a reason to have it be a subcommand, and I think it'd actually complicate things more than anything. So I've changed it to be the `-cov`/`--coverage` option again.
2. At the moment, a single run with the coverage option will get you both the verification results and the coverage results. I don't know why you think users will have to run Kani twice for an IDE, although that's what I expect to have in most CI jobs (i.e., run standard verification to decide whether to block or not a PR, then run coverage to maybe block if coverage percentage decreased too much or just to have reports available to reviewers).
3. Sorry, I've neglected IDE integration. Let me give you a more elaborate reply on your other comment about IDE integration.

Reply:
Made a few updates w.r.t. this comment:
 1. `--coverage` is an option in the standard Kani verification workflow. `cov` is a subcommand that aliases `kani-cov`. Also made clearer how `.kaniraw` and `.kanimap` files are produced and used in the coverage workflow.
 2. Added a subsection to discuss Kani VS Code integration. It's not very concrete since it'll likely just use the tooling we design here.